File: packages/sdk-python/src/aex_sdk/captive.py
"""Captive-portal detection via three standard probe endpoints.

Python mirror of the Rust ``aex-net::captive`` module; behaviour matches
``detect_network_state`` in Rust to the precision described by
``docs/protocol-v1.md`` §5.3.

Consensus rules (first match wins):

- Any probe saw a redirect or login-page body → ``NetworkState.CAPTIVE_PORTAL``
- All three probes behaved as expected → ``NetworkState.DIRECT``
- All three probes failed to complete → ``NetworkState.UNKNOWN``
- Any other mix → ``NetworkState.LIMITED``
"""

from __future__ import annotations

import enum
from typing import Iterable, Optional

import httpx


APPLE_URL = "http://captive.apple.com/hotspot-detect.html"
GOOGLE_URL = "http://www.google.com/generate_204"
MS_URL = "http://www.msftncsi.com/ncsi.txt"

APPLE_EXPECTED_BODY_FRAGMENT = "Success"
MS_EXPECTED_BODY = "Microsoft NCSI"

PROBE_TIMEOUT = 5.0


class NetworkState(str, enum.Enum):
    """High-level network reachability state.

    Serialised as lowercase snake_case so the value stream-to the
    ``AEX_NETWORK_STATE=<value>`` stdout flag matches the Rust crate's
    ``NetworkState::as_stdout_value``.
    """

    DIRECT = "direct"
    CAPTIVE_PORTAL = "captive_portal"
    LIMITED = "limited"
    UNKNOWN = "unknown"


class _ProbeVerdict(str, enum.Enum):
    OK = "ok"
    CAPTIVE = "captive"
    UNEXPECTED = "unexpected"
    FAILED = "failed"


def detect_network_state(
    client: Optional[httpx.Client] = None,
    *,
    apple_url: str = APPLE_URL,
    google_url: str = GOOGLE_URL,
    ms_url: str = MS_URL,
) -> NetworkState:
    """Fire the three probes and return the consensus state.

    ``client`` is optional; if omitted a short-lived :class:`httpx.Client`
    is constructed with follow_redirects=False. Callers who want the
    probe traffic to go through DoH should pass a client built by
    :func:`aex_sdk.resolver.build_http_client`.
    """
    owns_client = client is None
    if client is None:
        client = httpx.Client(follow_redirects=False, timeout=PROBE_TIMEOUT)
    try:
        results = [
            _probe_apple(client, apple_url),
            _probe_google(client, google_url),
            _probe_ms(client, ms_url),
        ]
    finally:
        if owns_client:
            client.close()

    return _consensus(results)


def _consensus(results: Iterable[_ProbeVerdict]) -> NetworkState:
    results = list(results)
    if _ProbeVerdict.CAPTIVE in results:
        return NetworkState.CAPTIVE_PORTAL
    if all(v == _ProbeVerdict.OK for v in results):
        return NetworkState.DIRECT
    if all(v == _ProbeVerdict.FAILED for v in results):
        return NetworkState.UNKNOWN
    return NetworkState.LIMITED


def _probe_apple(client: httpx.Client, url: str) -> _ProbeVerdict:
    try:
        r = client.get(url, timeout=PROBE_TIMEOUT)
    except httpx.HTTPError:
        return _ProbeVerdict.FAILED
    if r.is_redirect:
        return _ProbeVerdict.CAPTIVE
    if not r.is_success:
        return _ProbeVerdict.UNEXPECTED
    return (
        _ProbeVerdict.OK
        if APPLE_EXPECTED_BODY_FRAGMENT in r.text
        else _ProbeVerdict.CAPTIVE
    )


def _probe_google(client: httpx.Client, url: str) -> _ProbeVerdict:
    try:
        r = client.get(url, timeout=PROBE_TIMEOUT)
    except httpx.HTTPError:
        return _ProbeVerdict.FAILED
    if r.is_redirect:
        return _ProbeVerdict.CAPTIVE
    if r.status_code == 204:
        return _ProbeVerdict.OK
    return _ProbeVerdict.UNEXPECTED


def _probe_ms(client: httpx.Client, url: str) -> _ProbeVerdict:
    try:
        r = client.get(url, timeout=PROBE_TIMEOUT)
    except httpx.HTTPError:
        return _ProbeVerdict.FAILED
    if r.is_redirect:
        return _ProbeVerdict.CAPTIVE
    if not r.is_success:
        return _ProbeVerdict.UNEXPECTED
    return (
        _ProbeVerdict.OK
        if r.text.strip() == MS_EXPECTED_BODY
        else _ProbeVerdict.CAPTIVE
    )

File: packages/sdk-typescript/src/captive.ts
/**
 * Captive-portal detection via three standard probe endpoints.
 *
 * TypeScript mirror of the Rust `aex-net::captive` and Python
 * `aex_sdk.captive` modules. Behaviour is identical to the precision
 * described by `docs/protocol-v1.md` §5.3.
 */

/** Apple captive-portal probe URL. */
export const APPLE_URL = "http://captive.apple.com/hotspot-detect.html";
/** Google 204 endpoint. */
export const GOOGLE_URL = "http://www.google.com/generate_204";
/** Microsoft NCSI probe URL. */
export const MS_URL = "http://www.msftncsi.com/ncsi.txt";

const APPLE_EXPECTED_BODY_FRAGMENT = "Success";
const MS_EXPECTED_BODY = "Microsoft NCSI";

const PROBE_TIMEOUT_MS = 5000;

/**
 * Network reachability state.
 *
 * Serialised via {@link networkStateToStdoutValue} into the
 * `AEX_NETWORK_STATE=<value>` stdout flag emitted by the data-plane
 * binary. Values must stay in sync with Rust + Python.
 */
export type NetworkState =
  | "direct"
  | "captive_portal"
  | "limited"
  | "unknown";

/** Canonical string emitted by the AEX_NETWORK_STATE=<value> stdout flag. */
export function networkStateToStdoutValue(state: NetworkState): string {
  return state;
}

type ProbeVerdict = "ok" | "captive" | "unexpected" | "failed";

export interface DetectNetworkStateOptions {
  fetch?: typeof globalThis.fetch;
  appleUrl?: string;
  googleUrl?: string;
  msUrl?: string;
  timeoutMs?: number;
}

/** Fire the three probes in parallel and return the consensus state. */
export async function detectNetworkState(
  options: DetectNetworkStateOptions = {},
): Promise<NetworkState> {
  const fetchImpl = options.fetch ?? globalThis.fetch.bind(globalThis);
  const timeoutMs = options.timeoutMs ?? PROBE_TIMEOUT_MS;
  const appleUrl = options.appleUrl ?? APPLE_URL;
  const googleUrl = options.googleUrl ?? GOOGLE_URL;
  const msUrl = options.msUrl ?? MS_URL;

  const [apple, google, ms] = await Promise.all([
    probeApple(fetchImpl, appleUrl, timeoutMs),
    probeGoogle(fetchImpl, googleUrl, timeoutMs),
    probeMs(fetchImpl, msUrl, timeoutMs),
  ]);

  return consensus([apple, google, ms]);
}

export function consensus(results: readonly ProbeVerdict[]): NetworkState {
  if (results.includes("captive")) return "captive_portal";
  if (results.every((v) => v === "ok")) return "direct";
  if (results.every((v) => v === "failed")) return "unknown";
  return "limited";
}

async function probeApple(
  fetchImpl: typeof globalThis.fetch,
  url: string,
  timeoutMs: number,
): Promise<ProbeVerdict> {
  const resp = await safeFetch(fetchImpl, url, timeoutMs, "manual");
  if (!resp) return "failed";
  if (resp.status >= 300 && resp.status < 400) return "captive";
  if (!resp.ok) return "unexpected";
  const body = await resp.text().catch(() => null);
  if (body === null) return "failed";
  return body.includes(APPLE_EXPECTED_BODY_FRAGMENT) ? "ok" : "captive";
}

async function probeGoogle(
  fetchImpl: typeof globalThis.fetch,
  url: string,
  timeoutMs: number,
): Promise<ProbeVerdict> {
  const resp = await safeFetch(fetchImpl, url, timeoutMs, "manual");
  if (!resp) return "failed";
  if (resp.status >= 300 && resp.status < 400) return "captive";
  if (resp.status === 204) return "ok";
  return "unexpected";
}

async function probeMs(
  fetchImpl: typeof globalThis.fetch,
  url: string,
  timeoutMs: number,
): Promise<ProbeVerdict> {
  const resp = await safeFetch(fetchImpl, url, timeoutMs, "manual");
  if (!resp) return "failed";
  if (resp.status >= 300 && resp.status < 400) return "captive";
  if (!resp.ok) return "unexpected";
  const body = await resp.text().catch(() => null);
  if (body === null) return "failed";
  return body.trim() === MS_EXPECTED_BODY ? "ok" : "captive";
}

async function safeFetch(
  fetchImpl: typeof globalThis.fetch,
  url: string,
  timeoutMs: number,
  redirect: "manual" | "error" | "follow",
): Promise<Response | null> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    return await fetchImpl(url, { signal: ctrl.signal, redirect });
  } catch {
    return null;
  } finally {
    clearTimeout(timer);
  }
}
